### PR TITLE
[runtime] Remove compile time gc dependencies from object.c.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mono/metadata/object.c [hunk 1]
```diff
@@ -1932,23 +1930,22 @@ mono_class_create_runtime_vtable (MonoDomain *domain, MonoClass *klass, MonoErro
...
+	 * the mempool containing the vtable is destroyed when the domain is
+	 * unloaded. An alternative might be to allocate vtables in the GC
+	 * heap, but this does not seem to work (it leads to crashes inside
+	 * libgc). If that approach is tried, two gc descriptors need to be
+	 * allocated for each class: one for the root domain, and one for all
+	 * other domains. The second descriptor should contain a bit for the
+	 * vtable field in MonoObject, since we can no longer assume the
+	 * vtable is reachable by other roots after the appdomain is unloaded.
+	 */
+	if (!mono_gc_is_moving () && domain != mono_get_root_domain () && !mono_dont_free_domains)
```
Comment (joncham) [inline mono/metadata/object.c]: Probably fine, but this really has nothing to do with moving GC. Just GC that needs object descriptors valid after domain unload. 

I see we are just using `mono_gc_is_moving` as proxy for SGen so maybe that's okay.

File: mono/metadata/object.c [hunk 2]
```diff
@@ -6361,31 +6358,31 @@ mono_value_box_checked (MonoDomain *domain, MonoClass *klass, gpointer value, Mo
 
 	size = size - sizeof (MonoObject);
 
-#ifdef HAVE_SGEN_GC
-	g_assert (size == mono_class_value_size (klass, NULL));
-	mono_gc_wbarrier_value_copy ((char *)res + sizeof (MonoObject), value, 1, klass);
-#else
+	if (mono_gc_is_moving ()) {
+		g_assert (size == mono_class_value_size (klass, NULL));
+		mono_gc_wbarrier_value_copy ((char *)res + sizeof (MonoObject), value, 1, klass);
```
Comment (joncham) [inline mono/metadata/object.c]: Any reason we can't just use `mono_gc_wbarrier_value_copy` for all cases here?

**Comment by joncham** [comment]
It's fine as quick change, but if there is really sgen specific behavior perhaps we could just have explicit runtime check of `mono_gc_is_sgen` rather than confusing with `mono_gc_is_moving` ?

**Comment by vargaz** [comment]
Will add that.

**Pull request merged**